### PR TITLE
[Python Target] CARLA Demo

### PR DESCRIPTION
**Pull request by jackyk02**
This PR provides the code and instructions for running the [CARLA simulator](https://carla.org/) with Lingua Franca. For detailed guidance and troubleshooting, please refer to the [README](https://github.com/jackyk02/playground-lingua-franca/blob/main/examples/Python/src/CARLA/README.md) file.

### CODE REVIEW

File: examples/Python/src/CARLA/README.md [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+### Possible Configurations of CARLA Server
+
+The configuration of time-step and synchrony, leads for different settings. Here is a brief summary on the possibilities.
+
+|                   | Fixed time-step                                                        | Variable time-step                 |
+| ----------------- | ---------------------------------------------------------------------- | ---------------------------------- |
+| Synchronous mode  | Client is in total control over the simulation and its information.    | Risk of non reliable simulations.  |
+| Asynchronous mode | Good time references for information. Server runs as fast as possible. | Non easily repeatable simulations. |
+
+- **Synchronous mode + variable time-step.** This is almost for sure a non-desirable state. Physics cannot run properly when the time-step is bigger than 0.1s and. If the server has to wait for the client to compute the steps, this is likely to happen. Simulation time and physics will not be in synchrony. The simulation will not be reliable.
```
Comment (petervdonovan) [inline examples/Python/src/CARLA/README.md:51]:
```suggestion
- **Synchronous mode + variable time-step.** This is almost for sure a non-desirable state. Physics cannot run properly when the time-step is bigger than 0.1s, and if the server has to wait for the client to compute the steps, this is likely to happen. Simulation time and physics will not be in synchrony. The simulation will not be reliable.
```

File: examples/Python/src/CARLA/README.md [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+- **Asynchronous mode + variable time-step.** This is the default CARLA state. Client and server are asynchronous. The simulation time flows according to the real time. Reenacting the simulation needs to take into account float-arithmetic error, and possible differences in time steps between servers.
```
Comment (petervdonovan) [inline examples/Python/src/CARLA/README.md:52]: I don't see how float-arithmetic error is relevant, as floating-point error is perfectly deterministic/repeatable

File: examples/Python/src/CARLA/README.md [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+- **Asynchronous mode + fixed time-step.** The server will run as fast as possible. The information retrieved will be easily related with an exact moment in the simulation. This configuration makes possible to simulate long periods of time in much less real time, if the server is fast enough.
+- **Synchronous mode + fixed time-step.** The client will rule the simulation. The time step will be fixed. The server will not compute the following step until the client sends a tick. This is the best mode when synchrony and precision is relevant. Especially when dealing with slow clients or different elements retrieving information.
```
Comment (petervdonovan) [inline examples/Python/src/CARLA/README.md:54]:
```suggestion
- **Synchronous mode + fixed time-step.** The client will rule the simulation. The time step will be fixed. The server will not compute the following step until the client sends a tick. This is the best mode when synchrony and precision is relevant, especially when dealing with slow clients or different elements retrieving information.
```

File: examples/Python/src/CARLA/carla_circle.lf [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+          Sets synchronous mode.
+          """
+          settings = self.world.get_settings()
+          settings.fixed_delta_seconds = 0.05
+          settings.synchronous_mode = synchronous_mode
+          self.world.apply_settings(settings)
+
+      def setup_car(self):
+          """
+          Spawns actor-vehicle to be controled.
```
Comment (petervdonovan) [inline examples/Python/src/CARLA/carla_circle.lf:72]:
```suggestion
          Spawns actor-vehicle to be controlled.

```

File: examples/Python/src/CARLA/carla_manual_control.lf [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+          """
+          Sets synchronous mode.
+          """
+          settings = self.world.get_settings()
+          settings.synchronous_mode = synchronous_mode
+          self.world.apply_settings(settings)
+
+      def setup_car(self):
+          """
+          Spawns actor-vehicle to be controled.
```
Comment (petervdonovan) [inline examples/Python/src/CARLA/carla_manual_control.lf:71]:
```suggestion
          Spawns actor-vehicle to be controlled.

```

File: examples/Python/src/CARLA/carla_circle.lf [hunk 6]
```diff
@@ -0,0 +1,242 @@
+target Python {
+  threading: True
+}
+
+preamble {=
+  import carla
```
Comment (petervdonovan) [inline examples/Python/src/CARLA/carla_circle.lf:6]: Nit: this big chunk of Python code might be clearer and easier to work with if it were factored out into a separate `.py` file, but this is not important and addressing it is totally optional

**Review by petervdonovan — COMMENTED** [review]
There appears to be some code duplication between `carla_circle.lf` and `carla_manual_control.lf`. You could fix that by factoring out the Python code into a separate file as illustrated [here](https://github.com/lf-lang/lingua-franca/blob/36b3d4aac96850ac11d91e4c7a9d00cf4cc6a9f6/test/Python/src/StructAsType.lf#L6).

As is usually the case with duplication, you can probably solve this using composition by creating a class that is parameterized by a procedure that implements the behavior that is different between the two implementations. But I am happy with whatever you do as long as we are not stuck maintaining two copies of mostly the same code

File: examples/Python/src/CARLA/README.md (hunk 2 above)
Comment (jackyk02) [reply examples/Python/src/CARLA/README.md:52]: Removed

**Comment by jackyk02** [comment]
Thank you for the suggestions! I've factored out the Python code into a separate file. Please let me know if any additional changes are needed :)

**Review by petervdonovan — APPROVED** [review]
Great! It looks like my comments have been addressed.

File: examples/Python/src/CARLA/src/carla_async.lf [hunk 7]
```diff
@@ -0,0 +1,72 @@
+target Python {
+  threading: True,
```
Comment (lhstrh) [inline examples/Python/src/CARLA/src/carla_async.lf:2]:
```suggestion
```

File: examples/Python/src/CARLA/src/carla_sync.lf [hunk 8]
```diff
@@ -0,0 +1,72 @@
+target Python {
+  threading: True,
```
Comment (lhstrh) [inline examples/Python/src/CARLA/src/carla_sync.lf:2]:
```suggestion
```

**Comment by lhstrh** [comment]
I see that this was merged, but `./examples/Python/src/CARLA/src/carla_sync.lf` fails to compile.

**Pull request merged**